qlib/rl/order_execution/simulator_qlib.py

```diff
@@ -2,3 +2,370 @@
+            trade_price=market_price,
+            trade_value=market_price * exec_vol,
+            position=self.position - np.cumsum(exec_vol),
+            ffr=exec_vol / order.amount,
+            pa=pa,
+        )
+
+    def _collect_single_order_metric(
+        self,
+        order: Order,
+        datetime: pd.Timestamp,
+        market_vol: np.ndarray,
+        market_price: np.ndarray,
+        amount: float,  # intended to trade such amount
+        exec_vol: np.ndarray,
+    ) -> SAOEMetrics:
+        assert len(market_vol) == len(market_price) == len(exec_vol)
+
+        if np.abs(np.sum(exec_vol)) < EPS:
+            exec_avg_price = 0.0
+        else:
+            exec_avg_price = cast(float, np.average(market_price, weights=exec_vol))  # could be nan
+            if hasattr(exec_avg_price, "item"):  # could be numpy scalar
+                exec_avg_price = exec_avg_price.item()  # type: ignore
+
+        exec_sum = exec_vol.sum()
+        return SAOEMetrics(
+            stock_id=order.stock_id,
+            datetime=datetime,
+            direction=order.direction,
+            market_volume=market_vol.sum(),
+            market_price=market_price.mean() if len(market_price) > 0 else np.nan,
+            amount=amount,
+            inner_amount=exec_sum,
+            deal_amount=exec_sum,  # in this simulator, there's no other restrictions
+            trade_price=exec_avg_price,
+            trade_value=float(np.sum(market_price * exec_vol)),
+            position=self.position - exec_sum,
+            ffr=float(exec_sum / order.amount),
+            pa=_price_advantage(exec_avg_price, self._twap_price, order.direction),
+        )
+
+
+class SingleAssetQlibSimulator(Simulator[Order, SAOEState, float]):
+    def __init__(
+        self,
+        order: Order,
+        time_per_step: str,
+        qlib_config: dict,
+        inner_executor_fn: Callable[[str, CommonInfrastructure], BaseExecutor],
+        exchange_config: ExchangeConfig,
+    ) -> None:
+        super().__init__(
+            initial=None,  # TODO
+        )
+
+        assert order.start_time.date() == order.end_time.date()
+
+        self._order = order
+        self._order_date = pd.Timestamp(order.start_time.date())
+        self._trade_range = TradeRangeByTime(order.start_time.time(), order.end_time.time())
+        self._qlib_config = qlib_config
+        self._inner_executor_fn = inner_executor_fn
+        self._exchange_config = exchange_config
+
+        self._time_per_step = time_per_step
+        self._ticks_per_step = _convert_tick_str_to_int(time_per_step)
+
+        self._executor: Optional[NestedExecutor] = None
+        self._collect_data_loop: Optional[Generator] = None
+
+        self._done = False
+
+        self._inner_strategy = DecomposedStrategy()
+
+        self.reset(self._order)
+
+    def reset(self, order: Order) -> None:
+        instrument = order.stock_id
+
+        init_qlib(self._qlib_config, instrument)
```

Comment:
By design, a new simulator is created for every new episode. Therefore `init_qlib` can be called every new order is arrived. Make sure this is your intention, because it might cause performance issues.

Reply:
Will come back to this issue in later PRs.